### PR TITLE
Add FlowStepper.Indexed, Add more possibilites for header & footer rendering

### PR DESCRIPTION
**Pull request by ofreyssinet-ledger**
2 things in this PR:
- FlowStepper.Indexed
- Add more possibilities for header & footer rendering

### 1. FlowStepper.Indexed
**TL;DR**: Wrapper around the FlowStepper to set the active step with a prop `activeKey: string` where each children has their own `itemKey: string` prop.

This allows for usages with a lot of steps where dealing with indices could be
painful and error prone (for instance inserting/removing a step somewhere would shift
the indices of the following steps and navigation would be impacted).

By using string identifiers (`itemKey`) for each step, it's more "human readable"
and less error prone to setup a navigation logic between steps.

https://user-images.githubusercontent.com/91890529/151400255-9fd75b3c-3869-403a-a417-67660e97d884.mov

### 2. Add more possibilities for header & footer rendering

It's possible to give a `header` and `footer` prop to each step child component and it will be rendered as header/footer whenever that step is active.
To keep it easily composable, there are two additional render props `renderHeader` and `renderFooter` that are used in that case (when header/footer is an array of ReactNode ) and wrap the rendered header/footer.

https://user-images.githubusercontent.com/91890529/151423327-704954da-f630-4f1d-aaec-f8fdea84951e.mov

### CODE REVIEW

File: packages/react/src/components/navigation/FlowStepper/index.tsx [hunk 1]
```diff
@@ -108,20 +125,41 @@ function FlowStepper<ExtraProps>({
     },
   );
 
+  const headerContents =
+    header && Array.isArray(header)
```
Comment (ofreyssinet-ledger) [inline packages/react/src/components/navigation/FlowStepper/index.tsx:129]: there is certainly a better way to do this

File: packages/react/src/components/navigation/FlowStepper/index.tsx [hunk 2]
```diff
@@ -42,11 +47,19 @@ export interface Props<ExtraProps> {
   /**
    * An optional header displayed above the stepper.
    */
-  header?: (props: InnerProps & ExtraProps) => React.ReactNode;
+  header?: ((props: InnerProps & ExtraProps) => React.ReactNode) | StepExtra[];
```
Comment (LFBarreto) [inline packages/react/src/components/navigation/FlowStepper/index.tsx:50]: i dont really like those mixed types especially when one is an array and the other not :thinking: 
it forces you to check for types all along the component.
i would suggest in those cases splitting this field in 2 seperate ones 
```js
header?: ((props: InnerProps & ExtraProps) => React.ReactNode);
stepHeaders?: StepExtra[];

...

const headerContents = header || stepHeaders.find(...);
...
```

File: packages/react/src/components/navigation/FlowStepper/index.tsx [hunk 3]
```diff
@@ -42,11 +47,19 @@ export interface Props<ExtraProps> {
...
+  /**
+   * Custom rendering function to wrap header (only used if header is not a render function)
+   */
+  renderHeader?: (args: InnerProps & ExtraProps & { children: React.ReactNode }) => React.ReactNode;
   /**
    * An optional footer displayed below the body.
    */
-  footer?: (props: InnerProps & ExtraProps) => React.ReactNode;
+  footer?: ((props: InnerProps & ExtraProps) => React.ReactNode) | StepExtra[];
```
Comment (LFBarreto) [inline packages/react/src/components/navigation/FlowStepper/index.tsx:58]: same issue here

File: packages/react/src/components/navigation/FlowStepper/index.tsx (hunk 2 above)
Comment (LFBarreto) [reply packages/react/src/components/navigation/FlowStepper/index.tsx:50]: or only use a list and check on length :shrug:  to not have to mix types

**Review by LFBarreto — CHANGES_REQUESTED** [review]

**Review by elbywan — COMMENTED** [review]
> 1. FlowStepper.Indexed

I really like this part 👍 💯!

> 2. Add more possibilities for header & footer rendering

I'm wondering instead of making `header` and `footer` potential arrays, how about adding them as props in the `Step` component directly so we can associate the header, footer and the contents of the step in a single place? 

We could make it so that if the `Step` contains a header/footer prop it would override the one from the stepper component, and IMO it would be easier to use than passing arrays (with potentially undefined elements if we want a custom footer/header for a single step for instance).

**Comment by ofreyssinet-ledger** [comment]
> We could make it so that if the Step contains a header/footer prop it would override the one from the stepper component, and IMO it would be easier to use than passing arrays (with potentially undefined elements if we want a custom footer/header for a single step for instance).

@elbywan very good point yeah i thought about the general idea but didn't really know how to do that. Let me see I'm gonna give it a try now 👍

**Review by elbywan — APPROVED** [review]
Nice work!! 👍 

I was just wondering if you could add something in the existing storybook markdown to document the new features?
